lib/net.c

```diff
@@ -720,10 +720,62 @@ get_if_back(struct net_config *net_conf)
 	ETH_RSS_NONFRAG_IPV6_OTHER | \
 	ETH_RSS_IPV6_EX)
 
+static int
+check_port_rss_key_update(struct gatekeeper_if *iface, uint16_t port_id,
+	bool *rss_key_updatable, struct rte_eth_rss_conf *rss_conf)
+{
+	struct rte_eth_rss_conf cur_rss_conf;
+	int ret = rte_eth_dev_rss_hash_update(port_id, rss_conf);
+	if (ret == 0)
+		return ret;
+
+	ret = rte_eth_dev_rss_hash_conf_get(port_id, &cur_rss_conf);
+	if (ret < 0) {
+		G_LOG(WARNING,
+			"net: the %s interface does not support a critical hardware capability: update and get the RSS hash configuration\n",
+			iface->name);
+		return ret;
+	}
+
+	/*
+	 * At least two ports with different RSS hash keys
+	 * cannot be updated.
+	 */
+	if (!*rss_key_updatable &&
+			(cur_rss_conf.rss_key == NULL ||
+			cur_rss_conf.rss_key_len !=
+			rss_conf->rss_key_len ||
+			memcmp(rss_conf->rss_key,
+			cur_rss_conf.rss_key,
+			rss_conf->rss_key_len) != 0))
+		return -EINVAL;
+
+	if (cur_rss_conf.rss_key_len > GATEKEEPER_RSS_KEY_LEN)
+		return -EINVAL;
+
+	/* Update the default RSS hash key. */
+	memset(default_rss_key, 0, RTE_DIM(default_rss_key));
+	rte_memcpy(default_rss_key, cur_rss_conf.rss_key,
+		cur_rss_conf.rss_key_len);
```

Comment:
If we are updating the RSS key of the device with `default_rss_key` and checking it updated, why are we copying the current key to `default_rss_key`?

Reply:
The code comes from the assumption that Gatekeeper will abort only when at least two NICs with different hash keys cannot be updated, in which we cannot keep the hash keys consistent amog all the NICs.

If only one NIC's hash key cannot be updated, we can save it as the default, and update all other NICs' hash key later on.

Reply:
Well, in this case, the code is not doing what you are saying it should. For example, consider a case in which only one NIC does not support updating its RSS key:
1. `rte_eth_dev_rss_hash_update()` fails on the NIC that cannot update its key, what implies that `check_port_rss_key_update()` fails as well. `check_port_rss_key_update()` failing disables RSS altogether.
2. Even fixing item 1 above, variable `port_conf` will have the key that all NICs must use after the loop in `check_port_offloads()`, but Gatekeeper will later fail to update the NICs with that key because there is a NIC that doesn't support updating its key; it doesn't matter that is the same key it has.

Reply:
1. if ``rte_eth_dev_rss_hash_update()`` fails on that NIC, the return value will be ``ret < 0``, which will not fail ``check_port_rss_key_update()`` directly, instead, it tries to obtain the RSS hash key of that NIC.

2. Good point! To avoid the failure, we may have to introduce another data structure for each port recording its updatable status. If we abort when one of the NICs cannot support update, the solution will be straightforward as you suggested in other comments.

Reply:
I think that failing if any port doesn't support updating its RSS key is the way to go because ENA, our original target, does not allow either one to update, or to get the RSS key. That is, we are adding complexity to our code for a NIC that we don't even know if it exists: a NIC that gives its RSS key, but doesn't allow one to update it.